File: src/main/java/company/vk/polis/task1/Message.java
package company.vk.polis.task1;

public record Message(Integer id, String text, Integer senderId, Long timestamp, MessageState state) implements JavaEntity {
    @Override
    public Integer getId() {
        return id;
    }

    @Override
    public boolean isValid() { return id != null && text != null && senderId != null && timestamp != null && state != null; }
}

File: src/main/java/company/vk/polis/task1/entity/Message.kt
package company.vk.polis.task1.entity

data class Message(override val id: Int, val text: String?, val senderId: Int?, val timestamp: Long?) : Entity {
    sealed interface State
    object READ : State
    object UNREAD : State
    @JvmInline
    value class DELETED(private val userId: Int) : State

    var state: State = UNREAD
        private set

    fun read() {
        state = READ
    }

    fun delete(userId: Int) {
        state = DELETED(userId)

        println("> Message [$id] was deleted by user=$userId")
    }
}
